source/extensions/filters/network/dynamic_modules/abi_impl.cc

```diff
@@ -0,0 +1,273 @@
+
+#include "source/extensions/dynamic_modules/abi.h"
+#include "source/extensions/filters/network/dynamic_modules/filter.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace DynamicModules {
+namespace NetworkFilters {
+
+namespace {
+
+Network::ConnectionCloseType
+toEnvoyCloseType(envoy_dynamic_module_type_network_connection_close_type close_type) {
+  switch (close_type) {
+  case envoy_dynamic_module_type_network_connection_close_type_FlushWrite:
+    return Network::ConnectionCloseType::FlushWrite;
+  case envoy_dynamic_module_type_network_connection_close_type_NoFlush:
+    return Network::ConnectionCloseType::NoFlush;
+  case envoy_dynamic_module_type_network_connection_close_type_FlushWriteAndDelay:
+    return Network::ConnectionCloseType::FlushWriteAndDelay;
+  case envoy_dynamic_module_type_network_connection_close_type_Abort:
+    return Network::ConnectionCloseType::Abort;
+  case envoy_dynamic_module_type_network_connection_close_type_AbortReset:
+    return Network::ConnectionCloseType::AbortReset;
+  }
+  return Network::ConnectionCloseType::NoFlush;
+}
+
+// Helper to fill buffer slices from a Buffer::Instance into a pre-allocated array.
+void fillBufferSlices(const Buffer::Instance& buffer,
+                      envoy_dynamic_module_type_envoy_buffer* result_buffer_vector) {
+  Buffer::RawSliceVector raw_slices = buffer.getRawSlices();
+  auto counter = 0;
+  for (const auto& slice : raw_slices) {
+    result_buffer_vector[counter].length = slice.len_;
+    result_buffer_vector[counter].ptr = static_cast<char*>(slice.mem_);
+    counter++;
+  }
+}
+
+} // namespace
+
+extern "C" {
+
+bool envoy_dynamic_module_callback_network_filter_get_read_buffer_slices_size(
+    envoy_dynamic_module_type_network_filter_envoy_ptr filter_envoy_ptr, size_t* size) {
+  auto* filter = static_cast<DynamicModuleNetworkFilter*>(filter_envoy_ptr);
+  Buffer::Instance* buffer = filter->currentReadBuffer();
+
+  if (buffer == nullptr) {
+    return false;
+  }
+  *size = buffer->getRawSlices(std::nullopt).size();
+  return true;
+}
+
+size_t envoy_dynamic_module_callback_network_filter_get_read_buffer_slices(
+    envoy_dynamic_module_type_network_filter_envoy_ptr filter_envoy_ptr,
+    envoy_dynamic_module_type_envoy_buffer* result_buffer_vector) {
+  auto* filter = static_cast<DynamicModuleNetworkFilter*>(filter_envoy_ptr);
+  Buffer::Instance* buffer = filter->currentReadBuffer();
+
+  if (buffer == nullptr) {
+    return 0;
+  }
+
+  fillBufferSlices(*buffer, result_buffer_vector);
+  return buffer->length();
+}
+
+bool envoy_dynamic_module_callback_network_filter_get_write_buffer_slices_size(
+    envoy_dynamic_module_type_network_filter_envoy_ptr filter_envoy_ptr, size_t* size) {
+  auto* filter = static_cast<DynamicModuleNetworkFilter*>(filter_envoy_ptr);
+  Buffer::Instance* buffer = filter->currentWriteBuffer();
+
+  if (buffer == nullptr) {
+    return false;
+  }
+  *size = buffer->getRawSlices(std::nullopt).size();
```

Comment:
this is not a comment for this PR but the ABI in general including HTTP: i think we should create a fast pass to return the vector size as well as directly store the chunks into the vector allocated in a dynamic module so that we could avoid getRawSlices being called twice -- i think it shouldn't be that hard to make change to the core